### ISSUE TITLE
[datasets] Split creation of path structure from actual download of images

### ISSUE THREAD

**Issue by blenzi** [pyronear/pyro-vision#26]
## 🚀 Feature

Split creation of path structure and files that define train/test splitting from actual download of images

## Motivation

In order to use an external folder with images and preserve the test/train split defined in OpenFire without having to manually create the path structure.

**Comment by frgfm** [comment]
To follow up on #28, let's continue the discussion on design specs here @blenzi to come up with a good implementation.

Beforehand, I am starting to believe we should use the existing split we have already computed and set this directly in the annotations. It will remove the need for the `valid_pct` argument (and thus #25). I'll make a separate PR for this.

**Feature specifications**
- ability to download a limited amount of samples (specified by user)
- ability to use an existing image folder (`pathlib.Path`) instead of downloading samples
- Folder hierarchy inside `root`:
```
OpenFire
|--images
    |--0001.jpg
    |--0002.jpg
    ....
|--annotations
    |--binary-classification.json
    |--object-detection.json
    ....
```
(removing `train.pt` and `test.pt`)

**Workflow**
Here is what I had in mind:
- default value of `img_folder` would be `root.joinpath('OpenFire', 'images')`

- if  `download=False`: the class will try to use existing saved extracts and images (throwing an error otherwise like previously)

- if `download=True`: 
1. folder hierarchy will be created if it does not exist yet.
_it will create `image_folder` if it doesn't exist (ie. if not specified)_
2. the JSON extract will be download if it does not exist yet
3. the actual images will be downloaded to `img_folder` if they don't exist already (if `num_samples` is provided, only the first `num_samples` will be downloaded)

- Whatever  the value of `download`:
4. the existence of the selected images and annotations will be verified (throwing a warning if the specified amount of samples has not been downloaded or is not available)

What do you think?

**Comment by blenzi** [comment]
Nice, I agree @frgfm! A few questions and suggestions:
 
> Beforehand, I am starting to believe we should use the existing split we have already computed and set this directly in the annotations. It will remove the need for the `valid_pct` argument (and thus #25). I'll make a separate PR for this.

- I agree it is better. Since the annotations will be changed, do you also plan to add the local filename ? If not and it will still correspond to the order of the URLs, this should probably be documented.
- The idea is still to have an `train` argument or how do you plan to have the splitting for torch?

> 
> **Feature specifications**
> 
>     * ability to download a limited amount of samples (specified by user)

- Another feature I thought could be useful is to limit the number of samples to quickly test e.g. a new training method. I don't know if this should depend on having `download = True`, a new folder, etc. Do you foresee this feature ?
- If the download is limited to the first `num_samples`, the splitting between train/test has to be random (just like in the existing annotation) to avoid having a major difference in the fraction. It is probably worth documenting it for when one wants to create/use new annotations.

> 
>     * ability to use an existing image folder (`pathlib.Path`) instead of downloading samples

If a folder is passed, `OpenFire/images` will be a link pointing to it ?

**Comment by frgfm** [comment]
- For the filenames, yes, I just made the change, I'll open the PR shortly
- The `train` argument should remain in my opinion, to explicitly say which subset the returned object should be
- Mmmh that could be dangerous as well. And the dataset is still rather small for a Deep Learning dataset, to test it, we can easily make a loop iterate only N times over the dataloader.
- Yes, I respected the pseudo-random split!
- Yes, it will reference `OpenFire/images` by default but dynamically reference another folder if specified
